File: android_study/src/main/java/com/example/vladislav/androidstudy/MyApplication.java
package com.example.vladislav.androidstudy;

import android.app.Application;
import android.content.res.Configuration;
import android.widget.Toast;

/**
 * Created by Влад on 18.11.2017.
 */

// One needs to set next row in an AndroidManifest.xml
// android:name="com.example.vladislav.androidstudy.MyApplication"
// in an <application> node
//
// to use this instance of Application.

// To read
// https://github.com/codepath/android_guides/wiki/Understanding-the-Android-Application-Class
    // https://stackoverflow.com/questions/18002227/why-extend-an-application-class
    // https://stackoverflow.com/questions/3826905/singletons-vs-application-context-in-android
    // https://developer.android.com/reference/android/app/Application.html
    // http://www.programering.com/a/MzM4MDNwATU.html

    // Normally, developers do not extend this class
public class MyApplication extends Application {

    // Called when the application is starting, before any other application objects have been created.
    // Overriding this method is totally optional!
    @Override
    public void onCreate() {
        super.onCreate();
        Toast.makeText(this, getResources().getText(R.string.application_created_text),
                Toast.LENGTH_SHORT).show();
    }

    // Called by the system when the device configuration changes while your component is running.
    // Overriding this method is totally optional!
    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
    }

    // This is called when the overall system is running low on memory,
    // and would like actively running processes to tighten their belts.
    // Overriding this method is totally optional!
    @Override
    public void onLowMemory() {
        super.onLowMemory();
        Toast.makeText(this, getResources().getText(R.string.low_memory_text),
                Toast.LENGTH_SHORT).show();
    }

}

File: android_study/src/main/java/com/example/vladislav/androidstudy/MyApplication.kt
package com.example.vladislav.androidstudy

import android.app.Application
import android.content.res.Configuration
import android.widget.Toast
import com.example.vladislav.androidstudy.kotlin.demo.android.contentprovider.phonecontacts.di.ContactsComponent
import com.example.vladislav.androidstudy.kotlin.demo.android.contentprovider.phonecontacts.di.DaggerContactsComponent

class MyApplication: Application() {

    lateinit var contactsComponent: ContactsComponent

    // // Reference to the application graph that is used across the whole app
    // val loginComponent: LoginComponent by lazy {
    //     // Creates an instance of AppComponent using its Factory constructor
    //     // We pass the applicationContext that will be used as Context in the graph
    //     DaggerLoginComponent.factory().create(DaggerLoginComponent_LoginDependenciesComponent.factory().create())
    // }

    // Called when the application is starting, before any other application objects have been created.
    // Overriding this method is totally optional!
    override fun onCreate() {
        super.onCreate()
        // This way one passes an application context to dagger graph
        contactsComponent = DaggerContactsComponent.factory().create(applicationContext)
        // appComponent = DaggerApplicationComponent.create();
        Toast.makeText(
            this, resources.getText(R.string.application_created_text),
            Toast.LENGTH_SHORT
        ).show()
    }

    // Called by the system when the device configuration changes while your component is running.
    // Overriding this method is optional!
    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
    }

    // This is called when the overall system is running low on memory,
    // and would like actively running processes to tighten their belts.
    // Overriding this method is totally optional!
    override fun onLowMemory() {
        super.onLowMemory()
        Toast.makeText(
            this, resources.getText(R.string.low_memory_text),
            Toast.LENGTH_SHORT
        ).show()
    }
}